File: DSA/Umang-Singhal/Day16_Ques1.c
#include <stdio.h>
#include <stdlib.h>
void postorderTraversal(int nodeIndex, int left[], int right[], int values[]);
void buildTree(int numNodes, int values[], int left[], int right[]);
void postorderTraversal(int nodeIndex, int left[], int right[], int values[]) {
    if (nodeIndex == -1) {
        return;
    }
    postorderTraversal(left[nodeIndex], left, right, values);
    postorderTraversal(right[nodeIndex], left, right, values);
    printf("%d ", values[nodeIndex]);
}
void buildTree(int numNodes, int values[], int left[], int right[]) {
    for (int i = 0; i < numNodes; i++) {
        printf("Enter value for node %d: ", i);
        scanf("%d", &values[i]);
        printf("Enter index of left child of node %d (-1 if no left child): ", i);
        scanf("%d", &left[i]);
        printf("Enter index of right child of node %d (-1 if no right child): ", i);
        scanf("%d", &right[i]);
    }
}
int main() {
    int numNodes;
    printf("Enter the number of nodes in the binary tree: ");
    scanf("%d", &numNodes);
    int values[numNodes];
    int left[numNodes];
    int right[numNodes];
    buildTree(numNodes, values, left, right);
    int rootIndex;
    printf("Enter the index of the root node: ");
    scanf("%d", &rootIndex);
    printf("Postorder Traversal: ");
    postorderTraversal(rootIndex, left, right, values);
    printf("\n");
    return 0;
}

File: DSA/Ananya_Gambhir/Day16_ques1.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
        left = null;
        right = null;
    }
}

public class Day16_ques1 {

    public void postorderTraversal(TreeNode root) {
        postorderHelper(root);
    }

    private void postorderHelper(TreeNode node) {
        if (node == null) {
            return;
        }
        postorderHelper(node.left);
        postorderHelper(node.right);
        System.out.print(node.val + " ");
    }

    public static TreeNode buildTree() {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter the root value: ");
        int rootValue = scanner.nextInt();
        TreeNode root = new TreeNode(rootValue);

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            System.out.println("Enter left child of " + current.val + " (or -1 for no child): ");
            int leftValue = scanner.nextInt();
            if (leftValue != -1) {
                current.left = new TreeNode(leftValue);
                queue.add(current.left);
            }

            System.out.println("Enter right child of " + current.val + " (or -1 for no child): ");
            int rightValue = scanner.nextInt();
            if (rightValue != -1) {
                current.right = new TreeNode(rightValue);
                queue.add(current.right);
            }
        }
        scanner.close();
        return root;
    }

    public static void main(String[] args) {
        TreeNode root = buildTree();

        BinaryTreePostorderTraversal traversal = new BinaryTreePostorderTraversal();
        System.out.print("Postorder Traversal: ");
        traversal.postorderTraversal(root);
    }
}
